src/App.tsx

```diff
@@ -1,24 +1,253 @@
+import { TodoList } from './components/TodoList';
+import { TodoFooter } from './components/TodoFooter';
+import { Notification } from './components/Notification';
+
+import { Todo } from './types/Todo';
+import { FilterBy } from './types/FilterBy';
+import { Errors } from './types/Errors';
 
-const USER_ID = 0;
+import { client } from './utils/fetchClient';
+import * as utils from './utils/utils';
+
+const USER_ID = 11246;
+const BASE_ADD_URL = '/todos';
+const URL_GET = `?userId=${USER_ID}`;
+const DELAY_ERROR = 3000;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [hasTodoLoaderIds, setHasTodoLoaderIds] = useState<number[]>([]);
+  const [filterBy, setFilterBy] = useState<FilterBy>(FilterBy.all);
+  const [error, setError] = useState<Errors | null>(null);
+
+  const addLoader = (id: number) => {
+    setHasTodoLoaderIds(currentIds => [...currentIds, id]);
+  };
+
+  const deleteLoader = (id: number) => {
+    setHasTodoLoaderIds(currentIds => [...currentIds].filter(
+      hasLoaderId => hasLoaderId !== id,
+    ));
+  };
+
+  const changeAndToggleTodo = (
+    id: number,
+    field: keyof Todo,
+    value: Todo[typeof field],
+  ) => {
+    const indexOfMutableTodo = todos.findIndex(todo => todo.id === id);
+
+    switch (field) {
+      case 'completed':
+        setTodos(currentTodos => {
+          const copyTodos = [...currentTodos];
+
+          copyTodos[indexOfMutableTodo].completed
+            = !value;
+
+          return copyTodos;
+        });
+        break;
+      case 'title':
+        setTodos(currentTodos => {
+          const copyTodos = [...currentTodos];
+
+          copyTodos[indexOfMutableTodo].title
+            = `${value}`;
+
+          return copyTodos;
+        });
+        break;
+      default:
+        break;
+    }
+  };
+
+  const applyFilter = (filter: FilterBy) => {
+    setFilterBy(filter);
+  };
+
+  const addTodo = async (title: string) => {
+    if (title) {
+      const newTodo = {
+        title,
+        userId: USER_ID,
+        completed: false,
+      };
+
+      setHasTodoLoaderIds([0]);
+      setTempTodo({
+        ...newTodo,
+        id: 0,
+      });
+
+      try {
+        const addedTodo = await client.post<Todo>(BASE_ADD_URL, newTodo);
+
+        setTodos(currentTodos => [
+          ...currentTodos,
+          addedTodo,
+        ]);
+      } catch (caughtError) {
+        setError(Errors.add);
+        throw caughtError;
+      } finally {
+        setTempTodo(null);
+        setHasTodoLoaderIds([]);
+      }
+    } else {
+      setError(Errors.emptyTitle);
+    }
+  };
+
+  const deleteTodo = async (id: number) => {
+    addLoader(id);
+
+    try {
+      await client.delete(`${BASE_ADD_URL}/${id}`);
+      setTodos(currentTodos => {
+        return [...currentTodos].filter(todo => todo.id !== id);
+      });
+    } catch (errorCought) {
+      if (!error) {
```

Comment:
i guess you may get rid of this check

Reply:
I've thought, it helps avoid to additional render (if we add or remove multiple todos) - we set error once

Reply:
It prevent multiple adding errors to state